pkg/chunk/cassandra/scanner.go

```diff
@@ -0,0 +1,95 @@
+package cassandra
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+)
+
+// scanBatch represents a batch of rows read from Cassandra.
+type scanBatch struct {
+	hash       []byte
+	rangeValue []byte
+	value      []byte
+}
+
+type IndexValidator struct {
+	s *StorageClient
+	o *ObjectClient
+}
+
+func NewIndexValidator(s *StorageClient, o *ObjectClient) *IndexValidator {
+	return &IndexValidator{s, o}
+}
+
+func (i *IndexValidator) IndexScan(ctx context.Context, indexTable string, chunkTable string, from model.Time, to model.Time, out chan string) error {
+	q := i.s.readSession.Query(fmt.Sprintf("SELECT hash, range, value FROM %s", indexTable))
+
+	iter := q.WithContext(ctx).Iter()
+	defer iter.Close()
+	scanner := iter.Scanner()
+
+	wg := &sync.WaitGroup{}
+	batchChan := make(chan scanBatch)
+
+	for n := 0; n < 16; n++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for b := range batchChan {
+				i.checkEntry(ctx, chunkTable, from, to, out, b)
+			}
+		}()
+	}
+
+	for scanner.Next() {
+		b := scanBatch{}
+		if err := scanner.Scan(&b.hash, &b.rangeValue, &b.value); err != nil {
+			return errors.WithStack(err)
+		}
+		batchChan <- b
+	}
+	close(batchChan)
+	wg.Wait()
+	return errors.WithStack(scanner.Err())
+}
+
+func (i *IndexValidator) checkEntry(ctx context.Context, chunkTable string, from model.Time, to model.Time, out chan string, entry scanBatch) {
+	chunkID, _, _, err := parseChunkTimeRangeValue(entry.rangeValue, entry.value)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	c, err := chunk.ParseExternalKey("fake", chunkID)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	if from > c.Through || c.From > to {
+		fmt.Printf("ignoring chunk %s, outside time range\n", chunkID)
+		return
+	}
+
+	var exists int
+	err = i.o.readSession.Query(
+		fmt.Sprintf("SELECT count(*) FROM %s WHERE hash = ?", chunkTable),
+		c.ExternalKey(),
+	).WithContext(ctx).Scan(&exists)
+
+	if err != nil {
+		fmt.Println(err)
```

Comment:
is `fmt.Println()` thread-safe? it doesn't say anything about thread-safety here:
https://golang.org/pkg/fmt/#Println